sentinel/commiter/interval.go

```diff
@@ -0,0 +1,98 @@
+package commiter
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/benbjohnson/clock"
+	"github.com/ipfs/go-cid"
+	"github.com/rs/zerolog/log"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/areknoster/public-distributed-commit-log/pdclpb"
+	"github.com/areknoster/public-distributed-commit-log/sentinel"
+	"github.com/areknoster/public-distributed-commit-log/storage"
+	"github.com/areknoster/public-distributed-commit-log/thead"
+)
+
+// IntervalCommitter adds commit at given intervals.
+type IntervalCommitter struct {
+	headManager    thead.Manager
+	messageStorage storage.MessageStorage
+	pinner         sentinel.Pinner
+
+	ticker *clock.Ticker
+
+	mu          sync.Mutex
+	uncommitted []cid.Cid
+}
+
+func NewIntervalCommitter(headManager thead.Manager, messageStorage storage.MessageStorage, pinner sentinel.Pinner,
+	ticker *clock.Ticker) *IntervalCommitter {
+	ic := &IntervalCommitter{
+		headManager:    headManager,
+		messageStorage: messageStorage,
+		pinner:         pinner,
+		ticker:         ticker,
+	}
+	go ic.run()
+	return ic
+}
+
+func (i *IntervalCommitter) run() {
+	for {
+		<-i.ticker.C
+		i.mu.Lock()
+		if err := i.commit(); err != nil {
+			log.Error().Err(err).Msg("committing messages")
+			i.mu.Unlock()
+			continue
+		}
+		i.uncommitted = []cid.Cid{}
+		i.mu.Unlock()
```

Comment:
locking is good enough for now, but let's keep in mind, that if creating commit halts adding new messages till it's done. If it would be an issue, we could create a channel with batches, that would avoid locking

Reply:
Good point and I'm aware of this, but forgot to write a comment.